New: `no-invalid-this` rule (fixes #2815)

mysticatea opened the pull request:
I wrote, but it became larger than the first code...
I think what this rule does might not be simple, but is still not complex: just checking the function node and its parent node, almost.

The main reason that becomes larger than the first is processes for options.

Now, this rule two options.

1. `surelyUndefinedOnly` - If this is `true`, this rule warns only cases that `this` is `undefined` absolutely. In order to do so, I need to check whether or not this is under strict mode.
2. `allowInCallback` - If this is given an array of function names, `this` turns valid in callbacks of the functions.

As a test, I validated the source code of ESLint by this rule, there are some warnings in callbacks of `forEach` and `sinon.test`. This rule warned nothing (excepts in callbacks), so this implementation works as I expected. Actually both `this` in callbacks is valid, but we can turn off the warnings by `allowInCallback` option easily.

  ```
  Validating JavaScript files

  lib/rule-context.js
    83:8  error  Unexpected `this`  no-invalid-this

  lib/rules/no-irregular-whitespace.js
    129:37  error  Unexpected `this`  no-invalid-this

  ✖ 2 problems (2 errors, 0 warnings)

  Validating JavaScript test files

  tests/lib/cli-engine.js
    834:35  error  Unexpected `this`  no-invalid-this

  tests/lib/traverse.js
    133:23  error  Unexpected `this`  no-invalid-this
    135:8   error  Unexpected `this`  no-invalid-this

  ✖ 3 problems (3 errors, 0 warnings)
  ```

michaelficarra commented:
`surelyUndefinedOnly` is a poor name, since this should also catch top-level `this` in a module context, which is not an undefined value, but a reference error. Also, I don't think this should be an option. It should be the default and `alsoCheckFunctions` or something should be the option.

michaelficarra commented on docs/rules/no-invalid-this.md:
```diff
@@ -0,0 +1,294 @@
+# Disallow `this` keyword on the outside of constructors/methods (no-invalid-this)
+
+To access `this` keyword on the outside of constructors/methods is dangerous.
+Under strict mode, `this` is `undefined`, so raises an `ReferenceError`.
```
No, in strict mode, it's an undefined value. At the top level in a module, it's a reference error.

mysticatea replied:
Please teach me.
As far as I understand, `this` keyword under modules is `undefined` because below:

- [this keyword](http://www.ecma-international.org/ecma-262/6.0/#sec-this-keyword-runtime-semantics-evaluation)
- [ResolveThisBinding()](http://www.ecma-international.org/ecma-262/6.0/#sec-resolvethisbinding)  
  1. Let `envRec` be [GetThisEnvironment()](http://www.ecma-international.org/ecma-262/6.0/#sec-getthisenvironment) (returns [ModuleEnvironment](http://www.ecma-international.org/ecma-262/6.0/#sec-module-environment-records)).
    - `GetThisEnvironment()` returns an environment that its `HasThisBinding()` is `true`.
    - [ModuleEnvironmentRecord.HasThisBinding()](http://www.ecma-international.org/ecma-262/6.0/#sec-module-environment-records-hasthisbinding) returns always `true`.
  2. Return `envRec.GetThisBinding()`.
    - [ModuleEnvironmentRecord.GetThisBinding()](http://www.ecma-international.org/ecma-262/6.0/#sec-module-environment-records-getthisbinding) returns always `undefined`.

nzakas commented:
I think allowInCallbacks isn't necessary - I dying see that as a good general exception.

If we stick just with warning on this in module scope and this in a strict mode function, is that good enough to start?

mysticatea commented:
> I think allowInCallbacks isn't necessary - I don't see that as a good general exception.

I'm not sure `allowInCallbacks` is a good way, but I know there are libraries that uses `this` in callbacks: Koa, sinon, mocha, ...
Well, I see, `/** @this */ foo(function() { this; });` works, so it might be OK even if `allowInCallbacks` is removed.

> If we stick just with warning on this in module scope and this in a strict mode function, is that good enough to start?

I think, I don't want to check whether or not this is under strict mode.

- To not check whether or not this is under strict mode is simpler than to check it.
- Even if it's under not strict mode, unexpected `this` accesses are dangerous. 

```js
function a() {
    function b() {
        "use strict";
        function c() {
            function d() {
            }
        }
    }

    function e() {
    }
}
```

nzakas commented:
Is it possible to determine strict mode using escope?

ilyavolodin commented:
I was just going to mention that. I think escope has a flag that marks the scope as strict.

mysticatea commented:
Oh, I got it!
Thank you, I had not known the flag.

OK, I agreed to check only under strict mode is good.

mysticatea commented:
`Scope.isStrict` seems to be `true` on Method Initializer Shorthand.

```js
var obj = {foo() { console.log(this); z(x => console.log(x, this)); }};
```

I'm not sure, but I think this is wrong behavior...

nzakas commented:
Yeah, method initializers are not strict mode by default. It's probably confusing that with constructor methods, which are strict mode by default.

mysticatea commented:
Thank you.

I fixed escope, and I'm waiting for new release of escope.

nzakas commented:
Thanks for the update!

mysticatea commented:
I updated.

I removed all options, so now this rule is warning on `this` keywords in module scope and `this` keywords in a strict mode function merely.

I might be scolded, I wrote tests with a queer way.
This rule has many test patterns and those have different result each environment.  So patterns have valid environment list and invalid environment list.  I thought this way makes easy to keep the coverage.

nzakas commented:
This looks good to me but would like another set of eyes on it.

ilyavolodin commented on docs/rules/no-invalid-this.md:
```diff
@@ -0,0 +1,221 @@
+# Disallow `this` keyword on the outside of constructors/methods (no-invalid-this)
+
+To access `this` keyword on the outside of constructors/methods is dangerous.
```
I think this needs to be re-written. I'm not sure I'm getting the full picture of what this rule is supposed to flag. As far as I understand the point of the rule to flag usage of `this` keyword outside of classes (es6) or extendable objects (es3/es5).

mysticatea replied:
Ah, yes, you are right.

ilyavolodin commented:
I think implementation looks fine, but I'm afraid this rule is going to have extreme number of false positives. As far as I can tell, it will flag usage of `this` inside jQuery callback function, or any of the ES5 array methods, which are totally valid.

mysticatea commented:
@ilyavolodin 

> but I'm afraid this rule is going to have extreme number of false positives.

I think so too. Actually this rule warns the source code of ESLint at `this` keywords in callbacks of `forEach()` and `sinon.test()`.

In case of `forEach()`, I can add checking better, because it is included in the language spec.
But in other cases, wrong warnings are reported. 

I had thought the following code is valid:

```js
/** @this */
sinon.test(function() {
  this;
});
```

So, I agreed, removing the `allowInCallbacks` option is good.
But this code might be invalid since #2944.

mysticatea commented:
I tried to support callbacks of `.forEach()`-like methods.
I searched methods which has `thisArg` in its arguments, in the language spec:

- `Reflect.apply`
- `(Array|%TypedArray%).from`
- `(Array|%TypedArray%).prototype.every`
- `(Array|%TypedArray%).prototype.filter`
- `(Array|%TypedArray%).prototype.find`
- `(Array|%TypedArray%).prototype.findIndex`
- `(Array|%TypedArray%|Set|Map).prototype.forEach`
- `(Array|%TypedArray%).prototype.map`
- `(Array|%TypedArray%).prototype.some`

Now this rule does not warn `this` keywords in callbacks of those methods if `thisArg` is given.
The diff is here: https://github.com/mysticatea/eslint/commit/99e3360f47d9de64ac2dff5687c521d3c7b247c3

As a result, if `sinon.test` is excluded, this rule does not warn the source code of ESLint........ oh, one was remained.  I guess [this](https://github.com/eslint/eslint/blob/master/lib/rules/no-irregular-whitespace.js#L129) is not a false positive.

----

Next, I checked:

```js
/** @this */
sinon.test(function() {
  this;
});
```

This was invalid differently from [past](https://github.com/eslint/eslint/pull/2857#issuecomment-116244180).
So we lost an easy way to allow `this` keywords in callbacks, so I want to revive the `allowInCallback` option.

ilyavolodin commented:
@mysticatea How about
```js
$('.something').click(function() {
  $(this).hide();
}
```

mysticatea commented:
@ilyavolodin `[2, {"allowInCallback": ["click"]}]` is to allow the pattern.

Umm, using of a regular expression might be better, such as 
`[2, {"allowInCallback": "\\$\\(.+\\).click$"}]`

nzakas commented:
I think this is getting a bit more complicated than it needs to be. I would not allow specifying exceptions, we can just let people disable the rule where they need to.

mysticatea commented:
OK, I think designs which use `this` in callbacks are disappearing for arrow functions in future.
This rule is off by default, and people can choose using or not.
So I would not add any option.

I update the document a bit. Please review.

ilyavolodin commented:
Documentation is much clearer now. Sorry, haven't had time to review the code.

nzakas commented:
Lgtm. Can you turn this rule on in our .eslintrc file? I like to have as many rules on as possible so we can test them.

mysticatea commented:
Sure.
I added `no-invalid-this` into our .eslintrc file.

And I found that the following notation does not work as expected:

```js
sinon.test(/** @this sinon.sandbox */ function() {
    // ....
});
```

Because `context.getJSDocComment(theFunctionExpressionNode)` returns `null`.

So I modified a bit: https://github.com/mysticatea/eslint/commit/b16a31783bf36b23ad9e1cb366252a00211cde0d

nzakas commented:
Lgtm
